code/controllers/subsystem/fluids.dm

```diff
@@ -0,0 +1,266 @@
+
+/**
+ * # Fluid Subsystem
+ *
+ * A subsystem that processes the propagation and effects of a particular fluid.
+ *
+ * Both fluid spread and effect processing are handled through a carousel system.
+ * Fluids being spread and fluids being processed are organized into buckets.
+ * Each fresh (non-resumed) fire one bucket of each is selected to be processed.
+ * These selected buckets are then fully processed.
+ * The next fresh fire selects the next bucket in each set for processing.
+ * If this would walk off the end of a carousel list we wrap back to the first element.
+ * This effectively makes each set a circular list, hence a carousel.
+ */
+SUBSYSTEM_DEF(fluids)
+	name = "Fluid"
+	wait = 0.1 SECONDS
+	flags = SS_BACKGROUND|SS_KEEP_TIMING
+	runlevels = RUNLEVEL_GAME|RUNLEVEL_POSTGAME
+
+	/// What this subsystem processes. Yes, it's called 'case' entirely because of 'switch() case: '.
+	var/case = NONE
+
+	// Fluid spread processing:
+	/// The amount of time before a fluid node is created and when it spreads.
+	var/spread_wait = 1 SECONDS
+	/// The number of buckets in the spread carousel.
+	var/tmp/num_spread_buckets
+	/// The set of buckets containing fluid nodes to spread.
+	var/list/spread_carousel
+	/// The index of the spread carousel bucket currently being processed.
+	var/spread_bucket_index
+	/// The set of fluid nodes we are currently processing spreading for.
+	var/list/currently_spreading
+	/// Whether the subsystem has resumed spreading fluid.
+	var/resumed_spreading
+
+	// Fluid effect processing:
+	/// The amount of time between effect processing ticks for each fluid node.
+	var/effect_wait = 1 SECONDS
+	/// The number of buckets in the effect carousel.
+	var/num_effect_buckets
+	/// The set of buckets containing fluid nodes to process effects for.
+	var/list/effect_carousel
+	/// The index of the currently processing bucket on the effect carousel.
+	var/effect_bucket_index
+	/// The set of fluid nodes we are currently processing effects for.
+	var/list/currently_processing
+	/// Whether the subsystem has resumed processing fluid effects.
+	var/resumed_effect_processing
+
+/datum/controller/subsystem/fluids/Initialize(start_timeofday)
+	initialize_waits()
+	if (case & SS_PROCESSES_SPREADING)
+		initialize_spread_handling()
+	if (case & SS_PROCESSES_EFFECTS)
+		initialize_effect_handling()
+	return ..()
+
+/**
+ * Initializes the subsystem waits.
+ *
+ * Ensures that the subsystem's fire wait evenly splits the spread and effect waits.
+ */
+/datum/controller/subsystem/fluids/proc/initialize_waits()
+	case = NONE
+	if (spread_wait > 0)
+		case |= SS_PROCESSES_SPREADING
+	if (effect_wait > 0)
+		case |= SS_PROCESSES_EFFECTS
+
+	switch (case)
+		if (SS_PROCESSES_SPREADING|SS_PROCESSES_EFFECTS)
+			wait = Gcd(wait, Gcd(spread_wait, effect_wait))
+		if (SS_PROCESSES_SPREADING)
+			if (spread_wait > wait)
+				spread_wait = wait * round(spread_wait / wait, 1)
+			else
+				wait = spread_wait
+			effect_wait = wait
```

Comment:
is this just floor? what are you doing here

Reply:
That entire section was simplified to a single Gcd(...) call.